subsys/bluetooth/mesh/friend.c

```diff
@@ -316,13 +322,124 @@ static struct net_buf *create_friend_pdu(struct bt_mesh_friend *frnd,
 	return buf;
 }
 
-static int encrypt_friend_pdu(struct bt_mesh_friend *frnd, struct net_buf *buf, bool master_cred)
+struct unseg_app_sdu_meta {
+	struct bt_mesh_net_rx net;
+	const u8_t *key;
+	struct bt_mesh_subnet *subnet;
+	bool is_dev_key;
+	u8_t aid;
+	u8_t *ad;
+};
+
+static int unseg_app_sdu_unpack(struct bt_mesh_friend *frnd,
+				struct net_buf *buf,
+				struct unseg_app_sdu_meta *meta)
+{
+	u16_t app_idx = FRIEND_ADV(buf)->app_idx;
+	int err;
+
+	meta->subnet = bt_mesh_subnet_get(frnd->net_idx);
+	meta->is_dev_key = (FRIEND_ADV(buf)->app_idx == BT_MESH_KEY_DEV);
+	bt_mesh_net_header_parse(&buf->b, &meta->net);
+	err = bt_mesh_app_key_get(meta->subnet, app_idx, &meta->key,
+				  &meta->aid);
+	if (err) {
+		return err;
+	}
+
+	if (BT_MESH_ADDR_IS_VIRTUAL(meta->net.ctx.recv_dst)) {
+		meta->ad = bt_mesh_label_uuid_get(meta->net.ctx.recv_dst);
+		if (!meta->ad) {
+			return -ENOENT;
+		}
+	} else {
+		meta->ad = NULL;
+	}
+
+	return 0;
+}
+
+static int unseg_app_sdu_decrypt(struct bt_mesh_friend *frnd,
+				 struct net_buf *buf,
+				 const struct unseg_app_sdu_meta *meta)
+{
+	struct net_buf_simple sdu = {
+		.len = buf->len - 14,
+		.data = &buf->data[10],
+		.__buf = &buf->data[10],
+		.size = buf->len - 10,
+	};
```

Comment:
That's some pretty wild `net_buf_simple` construction 😄 

I wonder if it'd make sense to add a generic `net_buf_simple` helper that initialises a net_buf_simple to point to an existing buffer?

Reply:
Haha, yeah, this felt pretty dirty, but I couldn't find another way to cleanly pull out the sdu without altering the buf. Storing the state and manipulating the length and data didn't really seem any cleaner. If you think this is a real use case that shouldn't be solved in some other way, I'll create a helper function :+1:

Reply:
Double underscode (__buf) means that you are dealing with internal variables and that you should really know what you are doing. So I would think that in this case maybe the best thing would be to create a function in net/buf for what you want to do. Because you should not have this dependency into your part of the code.